### PR TITLE
[FLINK-20648][coordination][k8s] Start leader election service before creating JobMaster

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

When creating JobMaster service, we may recover from savepoints and update the HA storage(e.g. ConfigMap, ZooKeeper). So leader election service needs to be started first. Also we should guarantee that ConfigMap has been created after `leaderElectionService#start`. Then We need to block the constructor of KubernetesLeaderElectionDriver until the ConfigMap exists.

**Note:** This PR only applies to branch 1.12. In the master branch, we have a more graceful fix. Refer FLINK-11719 for more information.

The new added ITCase `KubernetesHighAvailabilityRecoverFromSavepointITCase` needs to be picked to master branch. It should pass without any production code change. 

## Brief change log

This PR will have the following major changes.
* Start leader election service before creating `JobMasterService` in `JobManagerRunnerImpl`
* `grantLeadership` and `revokeLeadership` should be called after creating `JobMasterService` in `JobManagerRunnerImpl`
* Make constructor of `KubernetesLeaderElectionDriver` blocking until the leader ConfigMap is created

## Verifying this change

* Update some related tests to pass
* Add a new ITCase `KubernetesHighAvailabilityRecoverFromSavepointITCase` for recovering from savepoints when HA is enabled

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wangyang0918** [comment]
cc @tillrohrmann @xintongsong Could you please have a look?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionService.java [hunk 1]
```diff
@@ -53,7 +59,7 @@ public synchronized void start(LeaderContender contender) {
 		this.contender = contender;
 
 		if (hasLeadership) {
-			contender.grantLeadership(issuedLeaderSessionId);
+			CompletableFuture.runAsync(() -> contender.grantLeadership(issuedLeaderSessionId), leaderElectorExecutor);
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionService.java:62]: This reveals the same problem as `StandaloneLeaderElectionService` does.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	public MiniClusterWithClientResource miniClusterResource = new MiniClusterWithClientResource(
+		new MiniClusterResourceConfiguration.Builder()
+			.setConfiguration(getConfiguration())
+			.setNumberTaskManagers(1)
+			.setNumberSlotsPerTaskManager(1)
+			.build());
+
+	private FlinkKubeClient flinkKubeClient;
+
+	private ClusterClient clusterClient;
```
Comment (xintongsong) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java:97]: IDE is complaining about "raw use of parameterized class"
```suggestion
	private ClusterClient<?> clusterClient;
```

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks for opening the PR, @wangyang0918.
I think my biggest concern is that, the internal correctness of `JobManagerRunnerImpl` is now depending on how the external leader election service interact with it. Please see my other comments for details.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/StandaloneLeaderElectionService.java [hunk 3]
```diff
@@ -44,8 +50,12 @@ public void start(LeaderContender newContender) throws Exception {
...
 		contender = Preconditions.checkNotNull(newContender);
 
-		// directly grant leadership to the given contender
-		contender.grantLeadership(HighAvailabilityServices.DEFAULT_LEADER_ID);
+		// Grant leadership in a separate thread
+		leaderElectorExecutor = Executors.newSingleThreadExecutor(
+			new ExecutorThreadFactory("LeaderElectorExecutor"));
+		CompletableFuture.runAsync(
+			() -> contender.grantLeadership(HighAvailabilityServices.DEFAULT_LEADER_ID),
+			leaderElectorExecutor);
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/StandaloneLeaderElectionService.java]: Good suggestion. Using the `whenComplete` to ensure the executing order could help to manage the changes in a smaller scope.

**Review by xintongsong — APPROVED** [review]
LGTM

**Comment by xintongsong** [comment]
Merged in 2c92ea9a06fc13d8adedd4cd5737ab89cd0a7334

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java [hunk 4]
```diff
@@ -268,21 +276,27 @@ private void unregisterJobFromHighAvailability() {
...
-			if (shutdown) {
-				log.debug("JobManagerRunner cannot be granted leadership because it is already shut down.");
-				return;
-			}
-
-			leadershipOperation = leadershipOperation.thenCompose(
-				(ignored) -> {
-					synchronized (lock) {
-						return verifyJobSchedulingStatusAndStartJobManager(leaderSessionID);
+		jobMasterCreationFuture.whenComplete(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java:279]: Why do we need the `jobMasterCreationFuture`? Wouldn't it be enough to wait on the lock and check whether the `JobManagerRunnerImpl` has been properly initialized?

Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java:279]: It is possible that the `LeaderElectionService` call `grantLeadership` on the same thread that `LeaderElectionService#start` is called (e.g., `StandaloneLeaderElectionService` & `TestingLeaderElectionService`). In that case, the lock does not block the thread from entering `grantLeadership` when `JobManagerRunnerImpl` has not been properly initialized.

Please see this comment:
https://github.com/apache/flink/pull/14485#discussion_r548399979

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java:279]: Now I understand it. Thanks for the clarification @xintongsong.

**Pull request closed without merging**